kernel/cClosure.ml

```diff
@@ -1099,154 +657,639 @@ module FNativeEntries =
+      check_pair env;
+      { mark = mark Cstr KnownR; term = FApp(!fPair, [|!fint;!fint;e1;e2|]) }
+
+    let mkFloatIntPair env f i =
+      check_pair env;
+      check_float env;
+      { mark = mark Cstr KnownR; term = FApp(!fPair, [|!ffloat;!fint;f;i|]) }
+
+    let mkLt env =
+      check_cmp env;
+      !fLt
+
+    let mkEq env =
+      check_cmp env;
+      !fEq
+
+    let mkGt env =
+      check_cmp env;
+      !fGt
+
+    let mkFLt env =
+      check_f_cmp env;
+      !fFLt
+
+    let mkFEq env =
+      check_f_cmp env;
+      !fFEq
+
+    let mkFGt env =
+      check_f_cmp env;
+      !fFGt
+
+    let mkFNotComparable env =
+      check_f_cmp env;
+      !fFNotComparable
+
+    let mkPNormal env =
+      check_f_class env;
+      !fPNormal
+
+    let mkNNormal env =
+      check_f_class env;
+      !fNNormal
+
+    let mkPSubn env =
+      check_f_class env;
+      !fPSubn
+
+    let mkNSubn env =
+      check_f_class env;
+      !fNSubn
+
+    let mkPZero env =
+      check_f_class env;
+      !fPZero
+
+    let mkNZero env =
+      check_f_class env;
+      !fNZero
+
+    let mkPInf env =
+      check_f_class env;
+      !fPInf
+
+    let mkNInf env =
+      check_f_class env;
+      !fNInf
+
+    let mkNaN env =
+      check_f_class env;
+      !fNaN
+
+    let mkArray env u t ty =
+      check_array env;
+      { mark = mark Whnf KnownR; term = FArray (u,t,ty)}
+
+  end
+
+module FredNative = RedNative(FNativeEntries)
+
+(************************************************************************)
+
+(** Hand-unrolling of the map function to bypass the call to the generic array
+    allocation *)
+let mk_clos_vect env v = match v with
+| [||] -> [||]
+| [|v0|] -> [|mk_clos env v0|]
+| [|v0; v1|] -> [|mk_clos env v0; mk_clos env v1|]
+| [|v0; v1; v2|] -> [|mk_clos env v0; mk_clos env v1; mk_clos env v2|]
+| [|v0; v1; v2; v3|] ->
+  [|mk_clos env v0; mk_clos env v1; mk_clos env v2; mk_clos env v3|]
+| v -> Array.Fun1.map mk_clos env v
+
+let ref_value_cache ({ i_cache = cache; _ }) tab ref =
+  try
+    KeyTable.find tab ref
+  with Not_found ->
+    let v =
+      try
+        let body =
+          match ref with
+          | RelKey n ->
+            let open! Context.Rel.Declaration in
+            let i = n - 1 in
+            let (d, _) =
+              try Range.get cache.i_env.env_rel_context.env_rel_map i
+              with Invalid_argument _ -> raise Not_found
+            in
+            begin match d with
+              | LocalAssum _ -> raise Not_found
+              | LocalDef (_, t, _) -> lift n t
+            end
+          | VarKey id -> assoc_defined id cache.i_env
+          | ConstKey cst -> constant_value_in cache.i_env cst
+        in
+        Def (inject body)
+      with
+      | NotEvaluableConst (IsPrimitive (u,op)) when Int.equal (CPrimitives.arity op) 0 ->
+        begin match FredNative.red_prim cache.i_env () op u [||] with
+          | Some m -> Def m
+          | None -> Primitive op
```

Comment:
When does a 0-arity primitive not reduce?

Reply:
I think it always reduces. Do you want an `assert false`?

Reply:
If it always reduces what's the point of it not being a regular definition?

Reply:
The point was to have only one source of truth for the max length, instead of hardcoding it once in ML and once in Gallina, with an axiom essentially saying that the two values agree. I can backtrack on it, though. I don't recall if this value can be architecture-dependent.

Reply:
We could have the primitive entry produce a regular definition, then the value doesn't appear in gallina.

Reply:
Good point, it would probably be less invasive than the current approach, and minimize code diffs.